src/python/WMCore/WebTools/RESTFormatter.py

```diff
@@ -66,26 +69,31 @@ def to_string(self, data):
             return self.json(data)
         if isinstance(data, dict) or isinstance(data, list):
             return json.dumps(data)
-        return str(data)
+        # Do not be misguided by the function name, this must return bytes, 
+        # otherwise the unit test
+        # testSupportedFormat (WMCore_t.WebTools_t.RESTFormat_t.RESTFormatTest)
+        # fails with "ValueError: Page handlers MUST return bytes. 
+        # Use tools.encode if you wish to return unicode."
+        return bytes(str(data), "utf-8")
```

Comment:
Honestly, I'm not sure whether `bytes` is really the data type that we need to return here. 
I would expect something like `data.encode('utf-8')` to be enough, which would still be a byte string but not preceded by `b'`, right? I'm not even sure this comment makes sense though.

I fear I won't ever have a full understanding of strings in python...

Reply:
#### .encode() vs bytes()

As you mentioned, `str(data).encode("utf-8")` would be the same as `bytes(str(data), "utf-8")` and we can use it if you prefer! They behave exactly the same in both python3 and modernized (with python future) python2


```python
> python3
Python 3.9.1 (default, Dec  8 2020, 07:51:42) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> data = {"key1": 2, "key2": 3.14}
>>> str(data).encode("utf-8")
b"{'key1': 2, 'key2': 3.14}"
>>> type(str(data).encode("utf-8"))
<class 'bytes'>
>>> bytes(str(data), "utf-8")
b"{'key1': 2, 'key2': 3.14}"
>>> type(bytes(str(data), "utf-8"))
<class 'bytes'>
```

```python
Python 2.7.13 (default, Sep 14 2017, 23:43:58) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from builtins import str, bytes
>>> data = {"key1": 2, "key2": 3.14}
>>> str(data).encode("utf-8")
b"{'key2': 3.14, 'key1': 2}"
>>> type(str(data).encode("utf-8"))
<class 'future.types.newbytes.newbytes'>
>>> bytes(str(data), "utf-8")
b"{'key2': 3.14, 'key1': 2}"
>>> type(bytes(str(data), "utf-8"))
<class 'future.types.newbytes.newbytes'>

```

I think that `b'` is only something that is used by the `__repr__` if the class  `bytes` to remind us that we are not dealing with a sequence of unicode but with a sequence of bytes.


#### do we really need to return bytes here?

When we use python2 interpreter, yes, it seems so. 

If we have `in WMCore/WebTools/RESTFormatter.RESTFormatter.to_string()`

```python
from builtins import str
[...]
    def to_string(self, data):
        if isinstance(data, GeneratorType):
            return self.json(data)
        if isinstance(data, dict) or isinstance(data, list):
            return json.dumps(data)
        return str(data)
```

then `python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python/WMCore_t/WebTools_t/RESTFormat_t.py` fails with

```plaintext
Traceback (most recent call last):
  File "/home/dmwm/unittestdeploy/wmagent/1.4.0-comp/sw/slc7_amd64_gcc630/external/py2-cherrypy/17.4.0-comp/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 630, in respond
    self._do_respond(path_info)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.0-comp/sw/slc7_amd64_gcc630/external/py2-cherrypy/17.4.0-comp/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 689, in _do_respond
    response.body = self.handler()
  File "/home/dmwm/unittestdeploy/wmagent/1.4.0-comp/sw/slc7_amd64_gcc630/external/py2-cherrypy/17.4.0-comp/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 794, in __set__
    raise ValueError(self.unicode_err)
ValueError: Page handlers MUST return bytes. Use tools.encode if you wish to return unicode.
```

It seems that cherrypy in python2 requires `to_string()` to return py2-`str` (which is bytes).

I could simply remove `from builtins import str`, but I assumed (hoped, bet) that cherrypy would require `to_string()` to return bytes also in python3, so I chaged it to

```python
from builtins import str
[...]
    def to_string(self, data):
        if isinstance(data, GeneratorType):
            return self.json(data)
        if isinstance(data, dict) or isinstance(data, list):
            return json.dumps(data)
        return bytes(str(data), "utf-8") # or also `str(data).encode("utf-8")`
```

This is an assumption, it may as weel be that cherrypy uses requires `to_string()` to return `str()` in both py2 (which means bytes) and py3 (which means unicode). I have no definitive answer now, we would need a py3 environment to test my  assumption, but cherrypy code have the same error in their py3 code

https://github.com/cherrypy/cherrypy/blob/c524a0da73a85ec4dd4ae6b1512c25854b19c8de/cherrypy/_cprequest.py#L779

So I would return bytes in `to_string()` also in py3.

Reply:
You're right. It's hard to be sure unless we have the python3 environment.
Do you think it would be helpful to add a # TODO, or # FIXME saying that this needs to be verified when running with python3? We have so many in the code, that maybe it won't make much difference...

Reply:
I fear the moment that we will be hit by a storm of errors and having a place to start looking for possible causes would be of much help. I think that such a flag would be useful! Or we can even have a new column in the python3 project in which we put possibly risky closed issues.
Or would you prefer to open new issues to keep track of these? (I fear that we would open a toooon of issues :sweat_smile:  )

Reply:
I think all these options would be good...perhaps the GH issue would have a "bigger impact" and easier to track(?) Please follow your taste here, anything works for me ;)

Reply:
Created #10222 to track this discussion